packages/flutter_tools/lib/src/dart/pub.dart

```diff
@@ -189,11 +191,32 @@ class _DefaultPub implements Pub {
     if (!dotPackages.existsSync()) {
       throwToolExit('$directory: pub did not create .packages file.');
     }
-
-    if (dotPackages.lastModifiedSync().isBefore(pubSpecYaml.lastModifiedSync())) {
-      throwToolExit('$directory: pub did not update .packages file '
-          '(pubspec.yaml timestamp: ${pubSpecYaml.lastModifiedSync()}; '
-          '.packages timestamp: ${dotPackages.lastModifiedSync()}).');
+    if (pubSpecYaml.lastModifiedSync() != originalPubspecYamlModificationTime) {
+      throwToolExit('$directory: unexpected concurrent modification of pubspec.yaml while running pub.');
+    }
+    // We don't check if dotPackages was actually modified, because as far as we can tell sometimes
+    // pub will decide it does not need to actually modify it.
+    // Since we rely on the file having a more recent timestamp, though, we do manually force the
+    // file to be more recently modified.
+    final DateTime now = DateTime.now();
+    if (now.isBefore(originalPubspecYamlModificationTime)) {
+      printError(
+        'Warning: File "${fs.path.absolute(pubSpecYaml.path)}" was created in the future. '
```

Comment:
Might want to include the unexpected future date in the error to give an extra hint.
Spooky! 👻

Reply:
oh good idea